File: sudoku/sudoku_solver.c
#include <igraph.h>
#include <iostream>
#include <vector>
#include <tuple>

using namespace std;

igraph_t create_graph() {
    igraph_t graph;
    igraph_empty(&graph, 0, IGRAPH_UNDIRECTED);

    // Step 1: Add all nodes
    vector<tuple<int, int, int, int, int, int, int, int, int>> nodes;
    for (int i = 0; i < 4; ++i) {
        for (int j = 0; j < 4; ++j) {
            for (int k = 0; k < 4; ++k) {
                for (int l = 0; l < 4; ++l) {
                    for (int m = 0; m < 4; ++m) {
                        for (int n = 0; n < 4; ++n) {
                            for (int o = 0; o < 4; ++o) {
                                for (int p = 0; p < 4; ++p) {
                                    for (int q = 0; q < 4; ++q) {
                                        tuple<int, int, int, int, int, int, int, int, int> node = make_tuple(i, j, k, l, m, n, o, p, q);
                                        nodes.push_back(node);
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    for (size_t i = 0; i < nodes.size(); ++i) {
        igraph_add_vertices(&graph, 1, 0);
    }

    // Step 2: Add all edges
    for (size_t i = 0; i < nodes.size(); ++i) {
        for (int r = 0; r < 9; ++r) {
            auto node = nodes[i];
            auto new_angles = node;
            get<r>(new_angles) = (get<r>(new_angles) + 1) % 4;
            auto it = find(nodes.begin(), nodes.end(), new_angles);
            if (it != nodes.end()) {
                int neighbor_index = distance(nodes.begin(), it);
                if (!igraph_are_connected(&graph, i, neighbor_index)) {
                    igraph_add_edge(&graph, i, neighbor_index);
                }
            }
        }
    }

    return graph;
}

void visualize_sudoku(igraph_t &graph, vector<int> node) {
    igraph_vector_t neighbors;
    igraph_vector_init(&neighbors, 0);

    int node_index = -1;
    for (int i = 0; i < igraph_vcount(&graph); ++i) {
        if (node == nodes[i]) {
            node_index = i;
            break;
        }
    }

    if (node_index != -1) {
        igraph_neighbors(&graph, &neighbors, node_index, IGRAPH_ALL);
        for (int i = 0; i < igraph_vector_size(&neighbors); ++i) {
            int neighbor_index = VECTOR(neighbors)[i];
            auto neighbor = nodes[neighbor_index];
            cout << "Neighbor " << i << ": ";
            for (int j = 0; j < 9; ++j) {
                cout << get<j>(neighbor) << " ";
            }
            cout << endl;
        }
    }

    igraph_vector_destroy(&neighbors);
}

int main() {
    igraph_t graph = create_graph();
    vector<int> node = {0, 1, 3, 2, 1, 3, 2, 1, 2};
    visualize_sudoku(graph, node);
    igraph_destroy(&graph);
    return 0;
}

File: sudoku/sudoku_solver.py
import networkx as nx


def create_graph():
    G = nx.Graph()
    
    # Step 1: Add all nodes
    nodes = []
    for i in range(4):
        for j in range(4):
            for k in range(4):
                for l in range(4):
                    for m in range(4):
                        for n in range(4):
                            for o in range(4):
                                for p in range(4):
                                    for q in range(4):
                                        print(i, j, k, l, m, n, o, p, q)
                                        node = (i, j, k, l, m, n, o, p, q)
                                        nodes.append(node)
    G.add_nodes_from(nodes)
    
    # Step 2: Add all edges
    for node in nodes:
        for r in range(9):
            print(node)
            new_angles = list(node)
            new_angles[r] = (new_angles[r] + 1) % 4
            neighbor = tuple(new_angles)
            if neighbor in nodes and not G.has_edge(node, neighbor):
                G.add_edge(node, neighbor, weight=1)
    
    return G


def solve_sudoku(start_sudoku):
    start = tuple(start_sudoku)
    end = tuple([0, 0, 0, 0, 0, 0, 0, 0, 1])
    path = nx.shortest_path(graph, source=start, target=end)
    return path

def visualize_sudoku(graph,node):

    node = tuple(node)
    neighbors = list(graph.neighbors(node))
    
    for i, neighbor in enumerate(neighbors):
        print(f"Neighbor {i}: {neighbor}")

if __name__ == '__main__':
    graph = create_graph()
    visualize_sudoku(graph, [0, 1, 3, 2, 1, 3, 2, 1, 2])
    # start_sudoku = [0, 0, 0, 0, 0, 3, 0, 0, 0]
    # path = solve_sudoku(start_sudoku)
    # print(path)